docs: update agent sections around auto-auth, caching, and templating (#15195)
hashicorp/vault: website/content/docs/agent/template.mdx

scenarios) and retry. On success, secrets defined in the templates will be
retrieved from Vault and rendered locally.

## Templating Language

The template output content can be provided directly as part of the `contents`

> option on a `template` stanza or as a separate `.ctmpl` file and specified in

Reviewer: ```suggestion
option in a `template` stanza or as a separate `.ctmpl` file and specified in
```

  template is rendered. The command will only run if the resulting template changes.
  The command must return within 30s (configurable), and it must have a successful
  exit code. Vault Agent is not a replacement for a process monitor or init system.
  This is deprecated in favor of the `exec` option.
- `command_timeout` `(duration: 30s)` - This is the maximum amount of time to
  wait for the optional command to return. This is deprecated in favor of the
  `exec` option.
- `error_on_missing_key` `(bool: false)` - Exit with an error when accessing
  a struct or map field/key that does notexist. The default behavior will print `<no value>`
  when accessing a field that does not exist. It is highly recommended you set this
  to "true".

> - `exec` `(object: optional)` - This is the optional exec block to give a command

Reviewer 2: Attempting to provide a bit more clarity without changing the intent of the content.

Suggestion:
The `exec` block gives a command when the template is rendered. However, the command will only run if the resulting template changes. Please note the `exec` block is *optional*, and the command within the `exec` block must run within 30s. The time for the command to run is configurable. The command must contain a successful exit code. Additionally, the parameters within the exec block just include the `command` string slice and the `timeout` string value.

Author: This was taken directly from the docs on the consul-template side: https://github.com/hashicorp/consul-template/blob/main/docs/configuration.md#templates. I can update it if it's clearer this way though.

Reviewer 2: Yes, please update based on my suggestion. The original on consul-template is pretty hard to read.

Author: Adopted this in [eb64fb9](https://github.com/hashicorp/vault/pull/15195/commits/eb64fb945c55953276443c88fcc87f1a662f00a9), with some minor tweaks.

  template is rendered. The command will only run if the resulting template changes.
  The command must return within 30s (configurable), and it must have a successful
  exit code. Vault Agent is not a replacement for a process monitor or init system.
  This is deprecated in favor of the `exec` option.
- `command_timeout` `(duration: 30s)` - This is the maximum amount of time to
  wait for the optional command to return. This is deprecated in favor of the
  `exec` option.
- `error_on_missing_key` `(bool: false)` - Exit with an error when accessing
  a struct or map field/key that does notexist. The default behavior will print `<no value>`
  when accessing a field that does not exist. It is highly recommended you set this
  to "true".
- `exec` `(object: optional)` - The exec block executes a command when the
  template is rendered. However, the command will only run if the resulting
  template changes. Please note the exec block is optional, and the command
  within the exec block must run within a configurable timeout which defaults

>   to 30s. The command must contain a successful exit code. Additionally,

Reviewer 3: Does this mean it needs to exit with code 0?

```suggestion
  to 30s. The command must have a successful exit code. Additionally,
```

>   to 30s. The command must contain a successful exit code. Additionally,

Reviewer: What happens if we get a non-zero exit code?

Author: The error is bubbled up the chain and it's treated as a render error.

Reviewer: So Agent will just log an error line essentially?

> - `exec` `(object: optional)` - The exec block executes a command when the
>   template is rendered. However, the command will only run if the resulting
>   template changes. Please note the exec block is optional, and the command
>   within the exec block must run within a configurable timeout which defaults
>   to 30s. The command must have a successful exit code. Additionally,
>   the parameters within the exec block just include the `command` string slice
>   and the `timeout` string value.

Reviewer: I think we can simplify the language here a bit, here's a rough attempt:

```suggestion
- `exec` `(object: optional)` - The exec block executes a command when the
  template is rendered and the output has changed. The block parameters are
  `command` `(string slice: required)` and `timeout` `(string: optional, defaults
  to 30s)`.
```

Author: Updated in [490ec79](https://github.com/hashicorp/vault/pull/15195/commits/490ec79e6364388f1205f2835cbaa2a7cdf4f68f).

>   to 30s. The command must contain a successful exit code. Additionally,
>   the parameters within the exec block just include the `command` string slice
>   and the `timeout` string value.

Reviewer: I found this last sentence quite confusing to read.

```suggestion
  to 30s. The command must contain a successful exit code. See [link]
  for details of how to configure this block.
```

Author: I'm not sure whether it's better to capture all of the info needed here or to direct users to yet another doc page, especially if we're pretty much copying the content over to this doc page. We have a similar pattern with the `wait` param further below.

This does not apply to certificates generated with `generate_lease: true`. If set
Vault Agent template will apply the non-renewable, leased secret rules.

> When Agent's auto-auth triggers a re-authentication, e.g. due to token expiry,

Reviewer 2: We avoid using *e.g.* as much as possible. 

Perhaps say: When Agent's auto-auth triggers a re-authentication, due to a token expiry for example, it generates a new token for Agent's use.

Author: Updated in [eb64fb9](https://github.com/hashicorp/vault/pull/15195/commits/eb64fb945c55953276443c88fcc87f1a662f00a9)

it generates a new token for Agent's use. This triggers a template server
restart, which fetched and re-renders a new set of certificates even if existing
certificates are valid.

## Templating Example

The following demonstrates configuring Vault Agent to template secrets using the

> AppRole Auth method:

Author: Would it be fine to update other examples to be like so as well? I stopped short of updating all the ones that I've found to be too long since I wasn't sure whether we'd want to keep a full config all the time.

> AppRole Auth method:

Reviewer 4: Go for it.  

If needed, we can add a note to say, "Refer to the [example configuration](/docs/agent#example-configuration) section for a complete Vault Agent configuration" or something like that. But short example looks cleaner.

> AppRole Auth method:

Author: I went ahead and shortened a few more example snippets. On these examples I've added a comment section at the top to indicate that there may be other config blocks like so:
```python
# Other Vault Agent configuration blocks
# ...
```

Let me know what you think.

> AppRole Auth method:

Reviewer 4: Nice!

This does not apply to certificates generated with `generate_lease: true`. If set
Vault Agent template will apply the non-renewable, leased secret rules.

> When Agent's auto-auth triggers a re-authentication, due to a token expiry for

Reviewer: ```suggestion
When Agent's auto-auth re-authenticates, due to a token expiry for
```

Reviewer: Maybe also worth explicitly saying what people would need to do to avoid this churn? i.e. give the auth method used by auto-auth a longer token TTL

Author: This is dependent on whether the token is a periodic or batch token as well (which disallows renewals entirely). Some users may not want their tokens to be long-lived. I don't think there's a way for us to consistently make this recommendation.

example, it generates a new token for Agent's use. This triggers a template

> server restart, which fetched and re-renders a new set of certificates even if

Reviewer 3: ```suggestion
server restart, which fetches and re-renders a new set of certificates even if
```

Reviewer 3: I also wonder if this paragraph should be called out in something like a Note text box for visibility.

Author: Changed it to be a Note box, though it's a bit lengthy. @taoism4504 thoughts?

  source      = "/tmp/agent/template.ctmpl"
  destination = "/tmp/agent/render.txt"
}

template {
  contents     = "{{ with secret \"secret/my-secret\" }}{{ .Data.data.foo }}{{ end }}"
  destination  = "/tmp/agent/render-content.txt"
}
```

> [consul-templating-language]: https://github.com/hashicorp/consul-template/blob/v0.27.1/docs/templating-language.md

Reviewer 3: Are we on consul-template 0.28.1 in Vault 1.10?
```suggestion
[consul-templating-language]: https://github.com/hashicorp/consul-template/blob/v0.28.1/docs/templating-language.md
```
